tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,450 @@
+
+		tb.timelyProposalsMu.Unlock()
+
+		return nil
+	default:
+		// TODO(nkryuchkov): handle other types
+		return nil
+	}
+}
+
+func (tb *TortoiseBeacon) handleVotingMessage(m VotingMessage) error {
+	epoch := m.Epoch()
+
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	mt := tb.classifyMessage(m, epoch)
+	switch mt {
+	case TimelyMessage:
+		tb.votesMu.Lock()
+		tb.votes[pair] = append(tb.votes[pair], m.Hash())
+		tb.votesMu.Unlock()
+
+		return tb.beaconCalculator.CountVotes(m) // TODO(nkryuchkov): count votes properly
+
+	default:
+		// TODO(nkryuchkov): handle other types
+		return nil
+	}
+}
+
+func (tb *TortoiseBeacon) handleBatchVotingMessage(m BatchVotingMessage) error {
+	epoch := m.Epoch()
+
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	pair := EpochRoundPair{
+		EpochID: epoch,
+		Round:   currentRound,
+	}
+
+	mt := tb.classifyMessage(m, epoch)
+	switch mt {
+	case TimelyMessage:
+		for _, hash := range m.HashList() {
+			tb.votesMu.Lock()
+			tb.votes[pair] = append(tb.votes[pair], hash)
+			tb.votesMu.Unlock()
+		}
+
+		return tb.beaconCalculator.CountBatchVotes(m) // TODO(nkryuchkov): count votes properly
+
+	default:
+		// TODO(nkryuchkov): handle other types
+		return nil
+	}
+}
+
+func (tb *TortoiseBeacon) classifyMessage(m Message, epoch types.EpochID) MessageType {
+	tb.currentRoundsMu.Lock()
+	currentRound := tb.currentRounds[epoch]
+	tb.currentRoundsMu.Unlock()
+
+	round := 0
+	if m, ok := m.(VotingMessage); ok {
+		round = m.Round()
+	}
+
+	switch {
+	case round <= currentRound-1:
+		return TimelyMessage
+	case round == currentRound-2:
+		return DelayedMessage
+	default:
+		return LateMessage
+	}
+}
+
+// for K rounds : in each round that lasts δ, wait for proposals to come in
+func (tb *TortoiseBeacon) roundTicker(epoch types.EpochID) {
+	ticker := time.NewTicker(tb.networkDelta)
+	defer ticker.Stop()
+
+	tb.currentRoundsMu.Lock()
+	tb.currentRounds[epoch] = 0
+	tb.currentRoundsMu.Unlock()
+
+	// round 0 already happened at this point, starting from round 1
+	// for next rounds wait for δ time, and construct a message that points to all messages from previous round received by δ
+	for i := 1; i < tb.lastPossibleRound(); i++ {
+		select {
+		case <-ticker.C:
+			tb.currentRoundsMu.Lock()
+			tb.currentRounds[epoch]++
+			round := tb.currentRounds[epoch]
+			tb.currentRoundsMu.Unlock()
+
+			if i == 1 {
+				// round 1, send hashed proposal
+				// create a voting message that references all seen proposals within δ time frame and send it
+				tb.timelyProposalsMu.RLock()
+				proposals := tb.timelyProposalsList[epoch]
+				tb.timelyProposalsMu.RUnlock()
+
+				for _, p := range proposals {
+					hash := hashATXList(p)
+
+					m := NewVotingMessage(epoch, round, hash)
+					tb.messageSender.Send(m)
```

Comment:
this will send multiple messages - why?

Reply:
Sending in one message is not pushed yet